### PR TITLE
feat(ds): Implements release boosting functionality for ds [TET-496]

### PR DESCRIPTION
**Pull request by ahmedetefy**
Sets releases that should be boosted with ds into the cache when a transaction is observed in the event manager. The logic is as follows once a transaction from a release that wasn't observed in the previous 24 hours is received, a cache key for that release is set with an expiration of one day and then that release is set into a list of boosted releases into the cache with an expiration of 1h, then the project config is invalidated so we recompute the project config with new dynamic sampling rule to boost that release with a hardcoded interval for one hour. If that release doesn't send any transactions in the next 24 hours i.e. after the 24 hour cache key expires and then starts sending transaction again, we want to start boosting the release again for an hour. This PR is one part of two parts, and only handles the setting of the cache and the invalidation of the project config, but does not include the dynamic sampling rules to be sent to relay. This is by design so we can merge this into production and monitor the performance impact of this logic before committing to adding the dynamic sampling rules

As a follow up, add a PR that only runs this logic if the feature flags for dynamic sampling are enabled, however we want to merge this without that check to monitor production load

### CODE REVIEW

File: tests/sentry/event_manager/test_event_manager.py [hunk 1]
```diff
@@ -2689,3 +2694,132 @@ def test_same_group_different_environment(self):
             last_seen=self.timestamp + 100,
             first_seen=self.timestamp + 100,
         )
+
+
+@region_silo_test
+class ReleaseDSLatestReleaseBoost(TestCase):
```
Comment (andriisoldatenko) [inline tests/sentry/event_manager/test_event_manager.py:2700]:
```suggestion
class DynamicSamplingLatestReleaseBoost(TestCase):
```

Comment (andriisoldatenko) [reply tests/sentry/event_manager/test_event_manager.py:2700]: 2 times Release

File: src/sentry/event_manager.py [hunk 2]
```diff
@@ -821,6 +823,18 @@ def _get_or_create_release_many(jobs, projects):
...
+                if data.get("type") == "transaction":
+                    try:
+                        release_observed_in_last_24h = observe_release(project_id, release.id)
+                        if not release_observed_in_last_24h:
+                            add_boosted_release(project_id, release.id)
+                            schedule_invalidate_project_config(
+                                project_id=project_id, trigger="dynamic_sampling:boost_release"
+                            )
+                    except Exception:
+                        pass
```
[the remark this answers is not in the document]
Comment (ahmedetefy) [reply src/sentry/event_manager.py:836]: Yeah that is a good idea

File: tests/sentry/event_manager/test_event_manager.py (hunk 1 above)
Comment (ahmedetefy) [reply tests/sentry/event_manager/test_event_manager.py:2700]: oops

**Review by iambriccardo — APPROVED** [review]

File: src/sentry/event_manager.py [hunk 3]
```diff
@@ -826,6 +828,24 @@ def _get_or_create_release_many(jobs, projects):
...
+                # Dynamic Sampling - Boosting latest release functionality
+                if (
+                    options.get("dynamic-sampling:boost-latest-release")
+                    and data.get("type") == "transaction"
+                ):
+                    with sentry_sdk.start_span(
+                        op="event_manager.dynamic_sampling_observe_latest_release"
+                    ), metrics.timer("event_manager.dynamic_sampling_observe_latest_release"):
+                        try:
+                            release_observed_in_last_24h = observe_release(project_id, release.id)
```
Comment (andriisoldatenko) [inline src/sentry/event_manager.py:840]: i know it's less important, but if we change `ONE_DAY_TIMEOUT` in `src/sentry/dynamic_sampling/latest_release_booster.py` this variable name will be not accurate. 

We can polish it later.

**Review by andriisoldatenko — APPROVED** [review]

File: mypy.ini [hunk 4]
```diff
@@ -44,7 +44,6 @@ files = fixtures/mypy-stubs,
         src/sentry/db/models/query.py,
         src/sentry/db/models/utils.py,
         src/sentry/digests/,
-        src/sentry/dynamic_sampling/,
```
[the remark this answers is not in the document]
Comment (ahmedetefy) [reply mypy.ini:47]: Yes, there is a PR coming up to fix this

[the remark this answers is not in the document]
Comment (ahmedetefy) [reply mypy.ini:47]: fwiw this is all new code, its not that we removed something that was already typed

[the remark this answers is not in the document]
Comment (ahmedetefy) [reply mypy.ini:47]: These are all new files that most of which landed in this last week, that we are still working on
I don't disagree with you that we should not turn typing off, it is just this one particular case where we needed to merge this to deploy it while someone from ops was around in our timezone, and with intention to fix soon thereafter

[the remark this answers is not in the document]
Comment (ahmedetefy) [reply mypy.ini:47]: https://github.com/getsentry/sentry/pull/40415/commits/cf484489e686d312ca4e4903c8afcd98e7d9da83

**Pull request merged**